ts_src/payments/p2sp.ts

```diff
@@ -0,0 +1,611 @@
+  if (!ecc.isPrivate(t_k)) throw new Error('shared-secret scalar is 0 or >= n');
+  return t_k;
+}
+
+/**
+ *  Calculate P_k
+ *  P_k = B_spend + t_k·G (compressed) -> x-only for P2TR
+ * @param spendPubKey - B_spend
+ * @param t_k
+ * @returns Pk
+ */
+export function calculateP_k(
+  spendPubKey: Uint8Array,
+  t_k: Uint8Array,
+): Uint8Array {
+  const Pk: Uint8Array | null = ecc.pointAddScalar(spendPubKey, t_k, true);
+  if (!Pk) throw new Error('pointAddScalar failed');
+  return Pk;
+}
+
+/**
+ * derive a silent payment output for a transaction.
+ * @param S - shared secret = (inputHash * B_scan) * a_sum   (compressed)
+ * @param spendPubkey - B_spend
+ * @param k - output number
+ * @returns {SilentOutput} the newly calculated output
+ */
+export function deriveSilentOutput(
+  S: Uint8Array,
+  spendPubkey: Uint8Array,
+  k: number,
+): SilentOutput {
+  // t_k = H_tag(SharedSecret, ser_P(S) || ser32BE(k))
+  const t_k: Uint8Array | null = calculateT_k(S, k);
+  if (!t_k) throw new Error('t_k: failed');
+
+  // P_k = B_spend + t_k·G (compressed) -> x-only for P2TR
+  const P_k: Uint8Array = calculateP_k(spendPubkey, t_k);
+  const P_xOnly: Uint8Array = toXOnly(P_k);
+  if (!P_xOnly) throw new Error('pointAddScalar failed');
+
+  return { pub_key: P_xOnly, tweak_key: t_k };
+}
+
+/**
+ * Label scalar: TaggedHash("BIP0352/Label", ser256(b_scan) || ser32BE(m))
+ * @param receiverScanPrivkey
+ * @param m
+ * @return tweaked label hash
+ */
+export function createLabelTweak(
+  receiverScanPrivkey: Uint8Array,
+  m: number,
+): Uint8Array {
+  // normalize m to uint32 per BIP
+  if (!Number.isInteger(m) || m < 0) throw new Error('label m must be uint32');
+  const m32 = m >>> 0;
+  const raw = taggedHash(
+    'BIP0352/Label',
+    tools.concat([receiverScanPrivkey, ser32BE(m32)]),
+  );
+  if (!ecc.isPrivate(raw)) throw new Error('label scalar is 0 or >= n');
+  return raw;
+}
+
+/**
+ * Generate labeled address Bm
+ *
+ * @param B_scan
+ * @param B_spend
+ * @param label
+ * @returns {L, Bm} - Label tweak and unencoded labeled address Bm
+ */
+export function generateLabelAndAddress(
+  B_scan: Uint8Array,
+  B_spend: Uint8Array,
+  label: number,
+): { L: Uint8Array; Bm: Uint8Array } {
+  // TaggedHash("BIP0352/Label", ser256(b_scan) || ser32BE(m))
+  const L: Uint8Array = createLabelTweak(B_scan, label);
+  // Bm = B_spend + hashBIP0352/Label(ser256(b_scan) || ser32(m))·G
+  const Bm: Uint8Array | null = ecc.pointAddScalar(B_spend, L, true);
+  if (!Bm) throw new Error('pointAddScalar(B_spend, L) failed');
+  return { L, Bm };
+}
+
+/**
+ * Scans a transaction's inputs and outputs to find any silent payments for the receiver.
+ * @param receiverScanPrivkey - b_scan
+ * @param receiverSpendPrivkey - b_spend
+ * @param inputHashTweak
+ * @param summedSenderPubkey - A_sum
+ * @param outputsToCheck - array of hex xOnly encoded outputs to check
+ * @param labelNonces
+ */
+export function scanForSilentPayments(
```

Comment:
May want to move this to a SilentPaymentTransaction class which allows you to process an entire transaction to see if there are any outputs for you